src/_pytest/python.py

```diff
@@ -704,66 +711,35 @@ def setup(self) -> None:
             func = partial(_call_with_optional_argument, teardown_module, init_mod)
             self.addfinalizer(func)
 
-    def _recurse(self, direntry: "os.DirEntry[str]") -> bool:
-        if direntry.name == "__pycache__":
-            return False
-        fspath = Path(direntry.path)
-        ihook = self.session.gethookproxy(fspath.parent)
-        if ihook.pytest_ignore_collect(collection_path=fspath, config=self.config):
-            return False
-        return True
-
-    def _collectfile(
-        self, fspath: Path, handle_dupes: bool = True
-    ) -> Sequence[nodes.Collector]:
-        assert (
-            fspath.is_file()
-        ), "{!r} is not a file (isdir={!r}, exists={!r}, islink={!r})".format(
-            fspath, fspath.is_dir(), fspath.exists(), fspath.is_symlink()
-        )
-        ihook = self.session.gethookproxy(fspath)
-        if not self.session.isinitpath(fspath):
-            if ihook.pytest_ignore_collect(collection_path=fspath, config=self.config):
-                return ()
-
-        if handle_dupes:
-            keepduplicates = self.config.getoption("keepduplicates")
-            if not keepduplicates:
-                duplicate_paths = self.config.pluginmanager._duplicatepaths
-                if fspath in duplicate_paths:
-                    return ()
-                else:
-                    duplicate_paths.add(fspath)
-
-        return ihook.pytest_collect_file(file_path=fspath, parent=self)  # type: ignore[no-any-return]
-
     def collect(self) -> Iterable[Union[nodes.Item, nodes.Collector]]:
-        # Always collect the __init__ first.
-        yield from self._collectfile(self.path / "__init__.py")
-
-        pkg_prefixes: Set[Path] = set()
-        for direntry in visit(self.path, recurse=self._recurse):
-            path = Path(direntry.path)
-
-            # Already handled above.
-            if direntry.is_file():
-                if direntry.name == "__init__.py" and path.parent == self.path:
+        # Always collect __init__.py first.
+        def sort_key(entry: "os.DirEntry[str]") -> object:
+            return (entry.name != "__init__.py", entry.name)
+
+        config = self.config
```

Comment:
I live by the [Rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) :grinning: 

But more seriously, I started out with this, with some common private function on `Directory`, but this stuff shouldn't be on `Directory` so I relented.

My goal is to simplify this loop such that it doesn't *need* deduplication because it's simple enough (which will also help plugins to implement it correctly). This things that pop out are:

- The `__pycache__` check can move to `pytest_ignore_collect` -- slight performance hit but probably minor.
- If we handle the `ihook` issue we can hoist it out.
- If `Path(direntry.path)` is not super slow we can hoist it out
- Maybe the `isinitpath` check can move to `pytest_ignore_collect` as well?

Reply:
Sounds good!